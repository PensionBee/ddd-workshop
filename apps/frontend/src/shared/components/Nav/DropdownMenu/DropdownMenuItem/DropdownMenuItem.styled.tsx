import { theme } from "@shared/styling/theme";
import styled, { css } from "styled-components";
import {
  enterFromLeft,
  enterFromRight,
  exitToLeft,
  exitToRight,
} from "../../Nav.styled";

type StyledListItemProps = {
  $isHighlighted?: boolean;
};

export const StyledListItem = styled.li<StyledListItemProps>`
  &:hover {
    background-color: ${() => theme.colors.grey6};
  }
  ${({ $isHighlighted }) =>
    $isHighlighted &&
    css`
      background-color: ${() => theme.colors.grey6};
    `}
`;

export const StyledNavLink = styled.a`
  width: 100%;
  display: flex;
  cursor: pointer;
  align-items: center;
  justify-content: space-between;
  padding: 16px 28px;
  &:focus-visible {
    outline: 3px solid ${() => theme.colors.blue100};
  }
`;

export const StyledSubNav = styled.div``;

export const StyledSubButtonStyles = css`
  width: 100%;
  display: flex;
  cursor: pointer;
  align-items: center;
  padding: 16px 28px;
  &:focus-visible {
    outline: 3px solid ${() => theme.colors.blue100};
  }
`;

export const StyledSubTriggerButton = styled.button`
  ${() => StyledSubButtonStyles}
  justify-content: space-between;
`;

export const StyledSubBackButton = styled.button`
  ${() => StyledSubButtonStyles}
  justify-content: flex-start;
  column-gap: 10px;
  border-bottom: 1px solid ${() => theme.colors.grey5};
`;

export const StyledSubContent = styled.div`
  z-index: 1;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  width: 100%;
  animation-duration: 250ms;
  animation-timing-function: ease;
  transform-origin: top center;
  background-color: white;
  overflow: hidden;
  box-shadow: 0px 5px 15px 0px rgb(0 0 0 / 18%);
  border-radius: 8px;
  transition: width, height, 300ms ease;
  @media only screen and (min-width: 600px) {
    width: auto;
  }
  &[data-motion="from-start"] {
    animation-name: ${enterFromLeft};
  }
  &[data-motion="from-end"] {
    animation-name: ${enterFromRight};
  }
  &[data-motion="to-start"] {
    animation-name: ${exitToLeft};
  }
  &[data-motion="to-end"] {
    animation-name: ${exitToRight};
  }
`;
