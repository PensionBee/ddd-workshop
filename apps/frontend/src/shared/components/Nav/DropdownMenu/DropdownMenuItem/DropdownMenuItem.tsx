import { LargeSystemIcon } from "@shared/components/Svg";
import { useLocation } from "react-router-dom";
import { Box } from "@shared/components/Box";
import type { NavMenuItem } from "../../Nav.types";
import { StyledListItem, StyledNavLink } from "./DropdownMenuItem.styled";

export const DropdownMenuItem: React.FC<NavMenuItem> = ({
  href,
  children,
  icon,
  ...props
}) => {
  const location = useLocation();

  return (
    <StyledListItem
      $isHighlighted={href === location.pathname}
      data-rc="DropdownMenuItem"
      data-highlighted={href === location.pathname}
      {...props}
    >
      <StyledNavLink href={href} tabIndex={0}>
        <Box d="flex" colGap={16} align="center">
          {icon ? <LargeSystemIcon w={24} h={24} icon={icon} /> : null}
          {children}
        </Box>
      </StyledNavLink>
    </StyledListItem>
  );
};
