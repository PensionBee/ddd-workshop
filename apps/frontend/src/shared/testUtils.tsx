import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
