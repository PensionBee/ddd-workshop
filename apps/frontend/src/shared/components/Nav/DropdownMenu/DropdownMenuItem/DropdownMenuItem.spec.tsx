import { routerRender, vi } from "@shared/testUtils";
import { DropdownMenuItem } from "./DropdownMenuItem";

const ITEM_CONTENT = "Item Content";
const EXAMPLE_LINK = "/example";

describe("DropdownMenuItem", () => {
  it("should render correctly", () => {
    const { getByText, asFragment } = routerRender(
      <DropdownMenuItem>{ITEM_CONTENT}</DropdownMenuItem>
    );

    expect(getByText(ITEM_CONTENT)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with href", () => {
    const { getByText, asFragment } = routerRender(
      <DropdownMenuItem href={EXAMPLE_LINK}>{ITEM_CONTENT}</DropdownMenuItem>
    );

    expect(getByText(ITEM_CONTENT).closest("a")).toHaveAttribute(
      "href",
      EXAMPLE_LINK
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should be highlighted if href matches pathname", () => {
    const { getByText, asFragment } = routerRender(
      <DropdownMenuItem href={EXAMPLE_LINK}>{ITEM_CONTENT}</DropdownMenuItem>,
      EXAMPLE_LINK // Sets current page route
    );

    expect(getByText(ITEM_CONTENT).closest("li")).toHaveAttribute(
      "data-highlighted"
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
