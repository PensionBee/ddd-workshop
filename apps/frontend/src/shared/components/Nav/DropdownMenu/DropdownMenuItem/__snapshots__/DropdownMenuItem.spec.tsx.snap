// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenuItem should be highlighted if href matches pathname 1`] = `
<DocumentFragment>
  <li
    class="sc-dkrFOg jrdWHL"
    data-highlighted="true"
    data-rc="DropdownMenuItem"
  >
    <a
      class="sc-hLBbgP jEcUor"
      href="/example"
      tabindex="0"
    >
      <div
        class="sc-gswNZR jRiSyp"
      >
        Item Content
      </div>
    </a>
  </li>
</DocumentFragment>
`;

exports[`DropdownMenuItem should render correctly 1`] = `
<DocumentFragment>
  <li
    class="sc-dkrFOg czKpsH"
    data-highlighted="false"
    data-rc="DropdownMenuItem"
  >
    <a
      class="sc-hLBbgP jEcUor"
      tabindex="0"
    >
      <div
        class="sc-gswNZR jRiSyp"
      >
        Item Content
      </div>
    </a>
  </li>
</DocumentFragment>
`;

exports[`DropdownMenuItem should render with href 1`] = `
<DocumentFragment>
  <li
    class="sc-dkrFOg czKpsH"
    data-highlighted="false"
    data-rc="DropdownMenuItem"
  >
    <a
      class="sc-hLBbgP jEcUor"
      href="/example"
      tabindex="0"
    >
      <div
        class="sc-gswNZR jRiSyp"
      >
        Item Content
      </div>
    </a>
  </li>
</DocumentFragment>
`;
